File: leetcode/contest-254/1969. Minimum Non-Zero Product of the Array Elements.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long LL;
LL P = 1e9 + 7;
LL qpow(LL a, LL b, LL p) {
    LL ans = 1ll % p;
    while (b) {
        if (b & 1) ans = 1ll * ans * a % p;
        a = 1ll * a * a % p;
        b >>= 1;
    }
    return ans;
}
class Solution {
   public:
    int minNonZeroProduct(int p) {
        LL t = (1ll << (p - 1)) - 1;
        LL a = (1ll << p) - 1, b = a - 1;
        // cout << b << " " << t << endl;
        return (a % P) * qpow(b % P, t, P) % P;
    }
};

File: leetcode/残酷刷题/1969. Minimum Non-Zero Product of the Array Elements.ts
function qpow (a: bigint, b: bigint, p: bigint) {
  let ans = 1n % p
  while (b) {
    if (b & 1n) ans = ans * a % p
    a = a * a % p
    b >>= 1n
  }
  return ans
}
function minNonZeroProduct (p: number): number {
  const P = BigInt(1e9 + 7); const p_b = BigInt(p)
  const a = 1n << p_b; const b = 1n << (p_b - 1n)
  const m1 = (a - 1n) % P; const m2 = (a - 2n) % P
  return Number(m1 * qpow(m2, b - 1n, P) % P)
};
